### PR TITLE
fix(cli/state rename): Update parent references on rename

### LINKED ISSUE #13179: Changing a parent resource's name causes panic
### What happened?

I wanted to change a resource's name which also has child resources but it caused a panic.

### Expected Behavior

successfully renames the resource and updates the child resource's parent reference

### Steps to reproduce

1. Create a custom resource with child resources
2. Change the parent resource's name using `pulumi state rename`

### Output of `pulumi about`

```bash
CLI          
Version      3.72.0
Go Version   go1.20.5
Go Compiler  gc

Plugins
NAME    VERSION
gcp     6.57.0
nodejs  unknown

Host     
OS       darwin
Version  13.4
Arch     arm64

This project is written in nodejs: executable='/Users/debkanchan/.nvm/versions/node/v16.15.0/bin/node' version='v16.15.0'

Current Stack: ride/infra/dev-2

TYPE                                                               URN
pulumi:pulumi:Stack                                                urn:pulumi:dev-2::infra::pulumi:pulumi:Stack::infra-dev-2
pulumi:providers:gcp                                               urn:pulumi:dev-2::infra::pulumi:providers:gcp::default_6_58_0
gcp:projects/service:Service                                       urn:pulumi:dev-2::infra::gcp:projects/service:Service::cloud-run-api
gcp:serviceAccount/account:Account                                 urn:pulumi:dev-2::infra::gcp:serviceAccount/account:Account::pulumi-service-account
gcp:projects/service:Service                                       urn:pulumi:dev-2::infra::gcp:projects/service:Service::artifact-registry-api
gcp:projects/service:Service                                       urn:pulumi:dev-2::infra::gcp:projects/service:Service::iam-api
gcp:projects/service:Service                                       urn:pulumi:dev-2::infra::gcp:projects/service:Service::secret-manager-api
gcp:serviceAccount/account:Account                                 urn:pulumi:dev-2::infra::gcp:serviceAccount/account:Account::artifact-registry-service-account
gcp:serviceAccount/account:Account                                 urn:pulumi:dev-2::infra::gcp:serviceAccount/account:Account::cloud-run-service-manager-service-account
gcp:artifactregistry/repository:Repository                         urn:pulumi:dev-2::infra::gcp:artifactregistry/repository:Repository::docker-registry
ride:infra:WorkloadIdentity                                        urn:pulumi:dev-2::infra::ride:infra:WorkloadIdentity::github-actions-workload-identity
ride:infra:WorkloadIdentity                                        urn:pulumi:dev-2::infra::ride:infra:WorkloadIdentity::pulumi-workload-identity
gcp:iam/workloadIdentityPool:WorkloadIdentityPool                  urn:pulumi:dev-2::infra::ride:infra:WorkloadIdentity$gcp:iam/workloadIdentityPool:WorkloadIdentityPool::pulumi-workload-identity-pool
gcp:iam/workloadIdentityPool:WorkloadIdentityPool                  urn:pulumi:dev-2::infra::ride:infra:WorkloadIdentity$gcp:iam/workloadIdentityPool:WorkloadIdentityPool::github-actions-workload-identity-pool
gcp:iam/workloadIdentityPoolProvider:WorkloadIdentityPoolProvider  urn:pulumi:dev-2::infra::ride:infra:WorkloadIdentity$gcp:iam/workloadIdentityPoolProvider:WorkloadIdentityPoolProvider::pulumi-oidc-provider
gcp:iam/workloadIdentityPoolProvider:WorkloadIdentityPoolProvider  urn:pulumi:dev-2::infra::ride:infra:WorkloadIdentity$gcp:iam/workloadIdentityPoolProvider:WorkloadIdentityPoolProvider::github-actions-oidc-provider
gcp:organizations/project:Project                                  urn:pulumi:dev-2::infra::gcp:organizations/project:Project::gcp-project
gcp:projects/service:Service                                       urn:pulumi:dev-2::infra::gcp:projects/service:Service::service-usage-api
ride:infra:FirebaseProject                                         urn:pulumi:dev-2::infra::ride:infra:FirebaseProject::dev-firebase
gcp:projects/service:Service                                       urn:pulumi:dev-2::infra::gcp:projects/service:Service::iam-service-account-credentials-api
gcp:projects/iAMMember:IAMMember                                   urn:pulumi:dev-2::infra::gcp:projects/iAMMember:IAMMember::artifact-registry-service-account-iam-member-2
gcp:projects/iAMMember:IAMMember                                   urn:pulumi:dev-2::infra::gcp:projects/iAMMember:IAMMember::artifact-registry-service-account-iam-member-3
gcp:projects/iAMMember:IAMMember                                   urn:pulumi:dev-2::infra::gcp:projects/iAMMember:IAMMember::artifact-registry-service-account-iam-member-0
gcp:projects/iAMMember:IAMMember                                   urn:pulumi:dev-2::infra::gcp:projects/iAMMember:IAMMember::artifact-registry-service-account-iam-member-1
gcp:projects/iAMMember:IAMMember                                   urn:pulumi:dev-2::infra::gcp:projects/iAMMember:IAMMember::pulumi-service-account-iam-member-0
gcp:projects/iAMMember:IAMMember                                   urn:pulumi:dev-2::infra::gcp:projects/iAMMember:IAMMember::github-cloud-run-service-account-iam-member-3
gcp:projects/iAMMember:IAMMember                                   urn:pulumi:dev-2::infra::gcp:projects/iAMMember:IAMMember::github-cloud-run-service-account-iam-member-4
gcp:projects/iAMMember:IAMMember                                   urn:pulumi:dev-2::infra::gcp:projects/iAMMember:IAMMember::github-cloud-run-service-account-iam-member-0
gcp:projects/iAMMember:IAMMember                                   urn:pulumi:dev-2::infra::gcp:projects/iAMMember:IAMMember::github-cloud-run-service-account-iam-member-1
gcp:projects/iAMMember:IAMMember                                   urn:pulumi:dev-2::infra::gcp:projects/iAMMember:IAMMember::github-cloud-run-service-account-iam-member-2
gcp:projects/service:Service                                       urn:pulumi:dev-2::infra::ride:infra:FirebaseProject$gcp:projects/service:Service::firestore-api
gcp:projects/service:Service                                       urn:pulumi:dev-2::infra::ride:infra:FirebaseProject$gcp:projects/service:Service::firebase-app-testers-api
gcp:projects/service:Service                                       urn:pulumi:dev-2::infra::ride:infra:FirebaseProject$gcp:projects/service:Service::firebase-management-api
gcp:projects/service:Service                                       urn:pulumi:dev-2::infra::ride:infra:FirebaseProject$gcp:projects/service:Service::firebase-database-api
gcp:firebase/project:Project                                       urn:pulumi:dev-2::infra::ride:infra:FirebaseProject$gcp:firebase/project:Project::firebase-project
gcp:firebase/databaseInstance:DatabaseInstance                     urn:pulumi:dev-2::infra::ride:infra:FirebaseProject$gcp:firebase/databaseInstance:DatabaseInstance::firebase-rtdb
gcp:firestore/database:Database                                    urn:pulumi:dev-2::infra::ride:infra:FirebaseProject$gcp:firestore/database:Database::firestore
gcp:serviceAccount/iAMBinding:IAMBinding                           urn:pulumi:dev-2::infra::ride:infra:WorkloadIdentity$gcp:serviceAccount/iAMBinding:IAMBinding::projects/ride-app-dev-2/serviceAccounts/artifact-registry-admin@ride-app-dev-2.iam.gserviceaccount.com-workload-identity-pool-iam-policy-binding
gcp:serviceAccount/iAMBinding:IAMBinding                           urn:pulumi:dev-2::infra::ride:infra:WorkloadIdentity$gcp:serviceAccount/iAMBinding:IAMBinding::projects/ride-app-dev-2/serviceAccounts/cloud-run-service-manager@ride-app-dev-2.iam.gserviceaccount.com-workload-identity-pool-iam-policy-binding
gcp:serviceAccount/iAMBinding:IAMBinding                           urn:pulumi:dev-2::infra::ride:infra:WorkloadIdentity$gcp:serviceAccount/iAMBinding:IAMBinding::projects/ride-app-dev-2/serviceAccounts/pulumi@ride-app-dev-2.iam.gserviceaccount.com-workload-identity-pool-iam-policy-binding


Found no pending operations associated with dev-2

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/DebkanchanSamadder
User           DebkanchanSamadder
Organizations  DebkanchanSamadder, ride

Pulumi locates its logs in /var/folders/7d/cln6swv17mq5fmjkpd6z6hnr0000gn/T/ by default
warning: Failed to get information about the Pulumi program's dependencies: failed to run "/Users/debkanchan/.nvm/versions/node/v16.15.0/bin/npm ls --json --depth=0": exit status 1
```

### Additional context

```bash
$ pulumi state rename urn:pulumi:dev-2::infra::ride:infra:FirebaseProject::dev-firebase firebase                               ✹

warning: This command will edit your stack's state directly. Confirm? Yes
================================================================================
The Pulumi CLI encountered a fatal error. This is a bug!
We would appreciate a report: https://github.com/pulumi/pulumi/issues/
Please provide all of the below text in your report.
================================================================================
Pulumi Version:   v3.72.0
Go Version:       go1.20.5
Go Compiler:      gc
Architecture:     arm64
Operating System: darwin
Panic:            fatal: error An assertion has failed: state edit produced an invalid snapshot. source error: child resource urn:pulumi:dev-2::infra::ride:infra:FirebaseProject$gcp:projects/service:Service::firestore-api refers to missing parent urn:pulumi:dev-2::infra::ride:infra:FirebaseProject::dev-firebase

goroutine 1 [running]:
runtime/debug.Stack()
        /opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/debug/stack.go:24 +0x64
main.panicHandler()
        /private/tmp/pulumi-20230615-4400-1iiccpl/pkg/cmd/pulumi/main.go:29 +0x34
panic({0x10653d5a0, 0x14000f91660})
        /opt/homebrew/Cellar/go/1.20.5/libexec/src/runtime/panic.go:884 +0x204
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
        /private/tmp/pulumi-20230615-4400-1iiccpl/sdk/go/common/util/contract/failfast.go:23
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.AssertNoErrorf({0x1069ac9a0, 0x14000f91630}, {0x105e828cf?, 0x0?}, {0x0?, 0x0?, 0x0?})
        /private/tmp/pulumi-20230615-4400-1iiccpl/sdk/go/common/util/contract/assert.go:51 +0x10c
main.totalStateEdit({0x1069c9fa0, 0x14000128010}, {0x1069db3f8, 0x14000d9e460}, 0x1, {{0x105e21818, 0x6}, 0x0, 0x0, 0x0, ...}, ...)
        /private/tmp/pulumi-20230615-4400-1iiccpl/pkg/cmd/pulumi/state.go:177 +0x290
main.runTotalStateEdit({0x1069c9fa0, 0x14000128010}, {0x0, 0x0}, 0x0?, 0x0?)
        /private/tmp/pulumi-20230615-4400-1iiccpl/pkg/cmd/pulumi/state.go:141 +0x13c
main.newStateRenameCommand.func1(0x0?, {0x1400049ff80, 0x2, 0x0?})
        /private/tmp/pulumi-20230615-4400-1iiccpl/pkg/cmd/pulumi/state_rename.go:144 +0x120
github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil.RunResultFunc.func1(0x14000c75200?, {0x1400049ff80, 0x2, 0x2})
        /private/tmp/pulumi-20230615-4400-1iiccpl/sdk/go/common/util/cmdutil/exit.go:112 +0x38
github.com/spf13/cobra.(*Command).execute(0x14000c75200, {0x1400049ff40, 0x2, 0x2})
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:944 +0x5b0
github.com/spf13/cobra.(*Command).ExecuteC(0x140007d2c00)
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x35c
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
main.main()
        /private/tmp/pulumi-20230615-4400-1iiccpl/pkg/cmd/pulumi/main.go:48 +0x3c
```

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by abhinav**
If a resource is renamed with `pulumi state rename`,
in addition to updating references in resources that refer to it
as a dependency,
also update references to the resource as a parent of another resource.

Testing:
Includes an integration test that reproduces the panic,
and a unit test that verifies the new behavior at a more isolated level.
To implement the integration test,
a new RunPulumiCommand method was added to ProgramTester
because we don't have other means of running `pulumi state rename`
with ProgramTest.

Resolves #13179

### CODE REVIEW

File: changelog/pending/20230616--cli-state--fix-panic-caused-by-an-invalid-stack-when-a-parent-resource-is-renamed-in-the-state.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: fix
+  scope: cli/state
+  description: Fix panic caused by an invalid stack when a parent resource is renamed in the state. Now, parent references are also updated when the resourcce is renamed.
```
Comment (justinvp) [inline changelog/pending/20230616--cli-state--fix-panic-caused-by-an-invalid-stack-when-a-parent-resource-is-renamed-in-the-state.yaml:4]:
```suggestion
  description: Fix panic caused by an invalid stack when a parent resource is renamed in the state. Now, parent references are also updated when the resource is renamed.
```

File: pkg/testing/integration/program.go [hunk 2]
```diff
@@ -962,6 +962,17 @@ func (pt *ProgramTester) runCommand(name string, args []string, wd string) error
 	return RunCommand(pt.t, name, args, wd, pt.opts)
 }
 
+// RunPulumiCommand runs a Pulumi command in the project directory.
+// For example:
+//
+//	pt.RunPulumiCommand("preview", "--stack", "dev")
+func (pt *ProgramTester) RunPulumiCommand(name string, args ...string) error {
+	// RunPulumiCommand uses 'name' for logging only.
```
Comment (justinvp) [inline pkg/testing/integration/program.go:970]: I was initially confused by this comment. I think you mean:

```suggestion
	// pt.runPulumiCommand uses 'name' for logging only.
```

**Review by justinvp — APPROVED** [review]

File: pkg/testing/integration/program.go (hunk 2 above)
Comment (abhinav) [reply pkg/testing/integration/program.go:970]: Oops, yes.

**Comment by abhinav** [comment]
Current dependencies on/for this PR:
* master
  * **PR #13190** <a href="https://app.graphite.dev/github/pr/pulumi/pulumi/13190" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/pulumi/pulumi/13190?utm_source=stack-comment).

**Comment by abhinav** [comment]
bors merge

**Comment by abhinav** [comment]
image

**Comment by justinvp** [comment]
bors retry

**Pull request merged**